File: packages/agentcore-strands/agent-container/container-sources/workspace_target.py
"""Workspace target validation shared by folder-addressed tools."""

from __future__ import annotations

import re
from dataclasses import dataclass
from typing import Iterable

TARGET_RE = re.compile(r"^[a-z0-9][a-z0-9-]{0,63}(/[a-z0-9][a-z0-9-]{0,63})*$")
RESERVED_SEGMENTS = {"memory", "skills"}
WORKSPACE_TARGET_DEPTH_CAP = 4


@dataclass(frozen=True)
class WorkspaceTargetResult:
    valid: bool
    normalized_path: str | None
    depth: int
    reason: str | None


def _normalize_route(route: str) -> str | None:
    trimmed = route.strip().rstrip("/")
    if not trimmed or trimmed == ".":
        return ""
    if not TARGET_RE.match(trimmed):
        return None
    if any(segment in RESERVED_SEGMENTS for segment in trimmed.split("/")):
        return None
    return trimmed


def parse_target(input: str, agents_md_routes: Iterable[str]) -> WorkspaceTargetResult:
    if input is None:
        return WorkspaceTargetResult(False, None, 0, "empty")

    trimmed = input.strip()
    if not trimmed:
        return WorkspaceTargetResult(False, None, 0, "empty")
    if trimmed == ".":
        return WorkspaceTargetResult(True, "", 0, None)
    if trimmed.startswith("/"):
        return WorkspaceTargetResult(False, None, 0, "absolute")
    if "\\" in trimmed or "?" in trimmed or "#" in trimmed:
        return WorkspaceTargetResult(False, None, 0, "malformed")
    if ".." in trimmed:
        return WorkspaceTargetResult(False, None, 0, "traversal")
    if not TARGET_RE.match(trimmed):
        return WorkspaceTargetResult(False, None, 0, "malformed")

    segments = trimmed.split("/")
    depth = len(segments)
    if any(segment in RESERVED_SEGMENTS for segment in segments):
        return WorkspaceTargetResult(False, None, depth, "reserved_name")
    if depth > WORKSPACE_TARGET_DEPTH_CAP:
        return WorkspaceTargetResult(False, None, depth, "depth_exceeded")

    routable = {
        normalized
        for route in agents_md_routes
        if (normalized := _normalize_route(route)) is not None
    }
    if trimmed not in routable:
        return WorkspaceTargetResult(False, None, depth, "not_routable")

    return WorkspaceTargetResult(True, trimmed, depth, None)

File: packages/api/src/lib/workspace-target.ts
export type WorkspaceTargetResult =
  | {
      valid: true;
      normalizedPath: string;
      depth: number;
      reason: null;
    }
  | {
      valid: false;
      normalizedPath: null;
      depth: number;
      reason:
        | "empty"
        | "absolute"
        | "traversal"
        | "malformed"
        | "reserved_name"
        | "depth_exceeded"
        | "not_routable";
    };

const TARGET_RE =
  /^[a-z0-9][a-z0-9-]{0,63}(\/[a-z0-9][a-z0-9-]{0,63})*$/;
const RESERVED_SEGMENTS = new Set(["memory", "skills"]);
export const WORKSPACE_TARGET_DEPTH_CAP = 4;

function normalizeRoute(route: string): string | null {
  const trimmed = route.trim().replace(/\/+$/g, "");
  if (!trimmed || trimmed === ".") return "";
  if (!TARGET_RE.test(trimmed)) return null;
  if (trimmed.split("/").some((segment) => RESERVED_SEGMENTS.has(segment))) {
    return null;
  }
  return trimmed;
}

export function parseWorkspaceTarget(
  input: string,
  agentsMdRoutes: string[],
): WorkspaceTargetResult {
  if (input == null) {
    return { valid: false, normalizedPath: null, depth: 0, reason: "empty" };
  }

  const trimmed = input.trim();
  if (!trimmed) {
    return { valid: false, normalizedPath: null, depth: 0, reason: "empty" };
  }
  if (trimmed === ".") {
    return { valid: true, normalizedPath: "", depth: 0, reason: null };
  }
  if (trimmed.startsWith("/")) {
    return { valid: false, normalizedPath: null, depth: 0, reason: "absolute" };
  }
  if (trimmed.includes("\\") || trimmed.includes("?") || trimmed.includes("#")) {
    return { valid: false, normalizedPath: null, depth: 0, reason: "malformed" };
  }
  if (trimmed.includes("..")) {
    return { valid: false, normalizedPath: null, depth: 0, reason: "traversal" };
  }
  if (!TARGET_RE.test(trimmed)) {
    return { valid: false, normalizedPath: null, depth: 0, reason: "malformed" };
  }

  const segments = trimmed.split("/");
  const depth = segments.length;
  if (segments.some((segment) => RESERVED_SEGMENTS.has(segment))) {
    return {
      valid: false,
      normalizedPath: null,
      depth,
      reason: "reserved_name",
    };
  }
  if (depth > WORKSPACE_TARGET_DEPTH_CAP) {
    return {
      valid: false,
      normalizedPath: null,
      depth,
      reason: "depth_exceeded",
    };
  }

  const routable = new Set(
    agentsMdRoutes
      .map(normalizeRoute)
      .filter((route): route is string => route !== null),
  );
  if (!routable.has(trimmed)) {
    return {
      valid: false,
      normalizedPath: null,
      depth,
      reason: "not_routable",
    };
  }

  return { valid: true, normalizedPath: trimmed, depth, reason: null };
}

